### PR TITLE
Add MPS/MPDO models with open and periodic boundary conditions

### LINKED ISSUE #726: rewrite MPS
Our MPS implementation has a quadratic complexity on the number of samples, which leads to crazy high memory requirements.
The reason is because a tensor to be contracted with `N_samples^2 x bond_dim^2` size is built and then reduced, instead of doing a loop over the tensors without storing the intermediate value. 
We should completely rewrite it.

### PR DESCRIPTION
**Pull request by LionSR**
After consulting with @PhilipVinc, I am adding MPS/MPDO models with open and periodic boundary conditions to netket. The implemented models should scale better than the previous MPSPeriodic model. 

Fix #726 

TODOs:

- [x] add more tests (right now NDM and MPDO are not covered)
- [x] add more documentation
- [x] add flags for checkpointing
- [x] add references (https://journals.aps.org/prx/pdf/10.1103/PhysRevX.8.011006, https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.100.040501, and shamelessly https://arxiv.org/abs/2401.14243 where we used MPDO)
- [x] make it convenient to extend to string-bond state and string-bond density operators using flax.nn.vmap (a revamp to #74, potentially in a different model repository). Added: right now the way it is should be natural.

Comments:
- I doubt anyone would use a DiagMPS so it was removed; however, a generalization of that, diagonal non-local string bond states (DiagSBS or nonlocal dSBS) have a close relationship with RBM as demonstrated in https://journals.aps.org/prx/pdf/10.1103/PhysRevX.8.011006. Features:
    - DiagSBS requires double the parameters compared to RBMs but doesn't necessitate complex parameters when there's a sign structure in the wavefunction.
    - In numerical benchmarks I've conducted, DiagSBS has shown nearly identical performance levels to RBM (sometimes even slightly more stable with SR) with the advantage of being natural to generalize when the local dimension d>2.
    - Maintaines should be easy, as the code is pretty short.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Thanks Sirui!

This looks quite good already. I'll go through it later today but in the meantime two quick comments:

 - Do add those new classes to [netket/docs/api/models.md](https://github.com/netket/netket/blob/master/docs/api/models.md) such that they get rendered in the documentation preview (and we can see if there are build errors there).
   - If you ever want to preview locally, you can install the `pip install -e 'netket[docs]'` dependencies locally and then `cd docs && make html && open _build/html/index.html` to preview them.

@gcarleo Do we want to scope MPS/MPDO in a `tensornetworks` submodule in `models`? Such that those models would be  constructed by doing `nk.models.tensornetworks.MPS(...)`?
I think this would create some order, what do you think?

File: netket/models/mpdo.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    Periodicity in the chain of MPDO tensors.
+    The chain of MPDO tensors is constructed as a sequence of identical
+    unit cells consisting of symperiod tensors. if None, symperiod equals the
+    number of physical degrees of freedom.
+    """
+    param_dtype: Any = jnp.float64
+    """complex or float, whether the variational parameters of the MPDO are real or complex."""
+    unroll: int = 1
+    """the number of scan iterations to unroll within a single iteration of a loop."""
+    kernel_init: NNInitFunc = default_kernel_init
```
Comment (PhilipVinc) [inline netket/models/mpdo.py:64]: Is this a good choice? Should we have a MPS specific initialiser?

Comment (wdphy16) [reply netket/models/mpdo.py:64]: For your information, ITensors uses [orthogonal init](https://github.com/ITensor/ITensors.jl/blob/9413240cc703a17abf6218951ba1171f3096802e/src/mps/mps.jl#L90) for `randomMPS` but I never compared it with Gaussian init

**Comment by LionSR** [comment]
Many thanks to both for the quick review! I missed the notifications during the week but I will address them early next week!

File: netket/models/mpdo.py (hunk 1 above)
Comment (LionSR) [reply netket/models/mpdo.py:64]: I did the plus identity matrix trick for both MPS + MPDO which was there for MPSPeriodic. I found that to be better than without indeed. Would the orthogonal initializer be better and easy to implement too?

File: netket/models/mpdo.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    The periodic boundary conditions imply that there are connections between the first and the last tensors.
+    """
+
+    hilbert: HomogeneousHilbert
+    """Hilbert space on which the state is defined."""
+    bond_dim: int
+    """Bond dimension of the MPDO tensors."""
+    kraus_dim: int = 2
+    """Kraus dimension of the MPDO tensors."""
+    symperiod: Optional[bool] = None
```
[the remark this answers is not in the document]
Comment (LionSR) [reply netket/models/mpdo.py:53]: symperiod was there with the previous MPSPeriodic

File: netket/models/mpdo.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    The periodic boundary conditions imply that there are connections between the first and the last tensors.
+    """
+
+    hilbert: HomogeneousHilbert
+    """Hilbert space on which the state is defined."""
+    bond_dim: int
+    """Bond dimension of the MPDO tensors."""
+    kraus_dim: int = 2
+    """Kraus dimension of the MPDO tensors."""
```
[the remark this answers is not in the document]
Comment (LionSR) [reply netket/models/mpdo.py]: added in the equations

File: netket/models/mpdo.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    def setup(self):
+        L, d, D, Χ = (
+            self.hilbert.size,
+            self.hilbert.local_size,
+            self.bond_dim,
+            self.kraus_dim,
+        )
+        self._L, self._d, self._D, self._Χ = L, d, D, Χ
+
+        self.param_dtype_cplx = dtype_complex(self.param_dtype)
```
[the remark this answers is not in the document]
Comment (LionSR) [reply netket/models/mpdo.py]: moved

**Comment by LionSR** [comment]
> Thanks Sirui! Looks good to me and I don't have anything urgent to say at the moment
> 
> There is still no mentioning of normalization or canonization, and if we only use MPS in MCMC then we indeed don't need it. It's an operation on the parameters, while the model is stateless, so I think when needed we can implement it as a utility function outside the model, and add it as a callback in the training driver
> 
> I don't know if people need the autoregressive sampling of MPS rather than the full contraction. If needed, I think it's not hard to port the code from my MPS-RNN on top of this

thanks, Dian! Your exact sampling code would certainly be very useful for MPS. I will leave it to you guys to decide what the best course of action would be.

File: netket/models/mpdo.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    r"""
+    A Matrix Product Density Operator (MPDO) with periodic boundary conditions for a quantum mixed state of discrete
+    degrees of freedom. The purification is used.
+
+    The MPDO is defined as
+
+    .. math::
+        \rho(s_1,\dots s_N, s_1',\dots s_N') = \sum_{\alpha_1, \dots, \alpha_{N-1}} A^{\alpha_1}[s_1, s_1'] \dots A^{\alpha_{N-1}}[s_N, s_N'] \dots A^{\alpha_N}[s_1, s_1'],
+
+    for arbitrary local quantum numbers :math:`s_i`, where :math:`A^{\alpha_i}[s_i, s_i']` is a tensor
```
[the remark this answers is not in the document]
Comment (LionSR) [reply netket/models/mpdo.py]: added

File: netket/models/mpdo.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    kraus_dim: int = 2
+    """Kraus dimension of the MPDO tensors."""
+    symperiod: Optional[bool] = None
+    """
+    Periodicity in the chain of MPDO tensors.
+    The chain of MPDO tensors is constructed as a sequence of identical
+    unit cells consisting of symperiod tensors. if None, symperiod equals the
+    number of physical degrees of freedom.
+    """
+    param_dtype: Any = jnp.float64
```
[the remark this answers is not in the document]
Comment (LionSR) [reply netket/models/mpdo.py]: changed everywhere

**Review by PhilipVinc — APPROVED** [review]
sorry for the delay! this looks great!
